### PR TITLE
G2 2023 v6 issue#13860 Implement test file for showDuggaservice.php

### PR DESCRIPTION
**Pull request by a21oscgu**
Implemented test file for showDuggaservice.php issue#13860

### CODE REVIEW

File: DuggaSys/tests/showDuggaservice_test.php [hunk 1]
```diff
@@ -0,0 +1,160 @@
+<?php
+
+include "../../Shared/test.php";
+
+$testsData = array(
+    'Get active users' => array(
+        'expected-output' => '{"debug":"[Guest] Missing hash\/password\/variant!","param":"{}","answer":"UNK","danswer":"UNK","score":0,"highscoremode":"","grade":"UNK","submitted":"","marked":"","deadline":"UNK","release":"UNK","files":[],"userfeedback":"UNK","feedbackquestion":"UNK","variant":"UNK","ishashindb":false,"variantsize":"UNK","variantvalue":"UNK","password":"UNK","hashvariant":"UNK","isFileSubmitted":"UNK","isTeacher":0,"variants":[],"duggaTitle":"UNK","hash":"UNK","hashpwd":"UNK","opt":"UNK","link":"UNK"}',
```
Comment (b21antbr) [inline DuggaSys/tests/showDuggaservice_test.php:7]: after expected out a "query-before-test" is needed to input the pre-values. Here an INSERT is needed
Pre-query: INSERT INTO groupdugga(hash,active_users)  VALUES ("hnf3j58s", 5);

Comment (b21antbr) [inline DuggaSys/tests/showDuggaservice_test.php:7]: after this is done you will need a "query-after-test" stating DELETE FROM groupdugga WHERE hash = “hnf3j58s”;

File: DuggaSys/tests/showDuggaservice_test.php [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+            'opt' => 'UPDATEAU',
+            'hash' => 'hnf3j58s'
+        )),
+        'filter-output' => serialize(array( // Filter what output to use in assert test, use none to use all ouput from service
+            'debug',
+            'readonly'
+        )),
+    ),
+    'Create active users' => array(
+        'expected-output' => '{"debug":"[Guest] Missing hash\/password\/variant!","param":"{}","answer":"UNK","danswer":"UNK","score":0,"highscoremode":"","grade":"UNK","submitted":"","marked":"","deadline":"UNK","release":"UNK","files":[],"userfeedback":"UNK","feedbackquestion":"UNK","variant":"UNK","ishashindb":false,"variantsize":"UNK","variantvalue":"UNK","password":"UNK","hashvariant":"UNK","isFileSubmitted":"UNK","isTeacher":0,"variants":[],"duggaTitle":"UNK","hash":"UNK","hashpwd":"UNK","opt":"UNK","link":"UNK"}',
```
Comment (b21antbr) [inline DuggaSys/tests/showDuggaservice_test.php:19]: a query-after-test is needed to remove the new user. DELETE FROM groupdugga WHERE hash = “tj7dh2nb”;

File: DuggaSys/tests/showDuggaservice_test.php [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        'service' => 'localhost/LenaSYS/DuggaSys/showDuggaservice.php',
+        'service-data' => serialize(array( // Data that service needs to execute function
+            'opt' => 'UPDATEAU',
+            'hash' => 'tj7dh2nb',
+            'AUtoken' => '999'
+        )),
+        'filter-output' => serialize(array( // Filter what output to use in assert test, use none to use all ouput from service
```
Comment (c21axepe) [inline DuggaSys/tests/showDuggaservice_test.php:26]: In filter all "labels" in expected output should be implemented.
Ex: 
"debug",
"param"

File: DuggaSys/tests/showDuggaservice_test.php [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+            'none'
+        )),
+    ),
+    'update active users' => array(
+        'expected-output' => '{"debug":"[Guest] Missing hash\/password\/variant!","param":"{}","answer":"UNK","danswer":"UNK","score":0,"highscoremode":"","grade":"UNK","submitted":"","marked":"","deadline":"UNK","release":"UNK","files":[],"userfeedback":"UNK","feedbackquestion":"UNK","variant":"UNK","ishashindb":false,"variantsize":"UNK","variantvalue":"UNK","password":"UNK","hashvariant":"UNK","isFileSubmitted":"UNK","isTeacher":0,"variants":[],"duggaTitle":"UNK","hash":"UNK","hashpwd":"UNK","opt":"UNK","link":"UNK"}',
```
Comment (b21antbr) [inline DuggaSys/tests/showDuggaservice_test.php:31]: a query-before-test is needed to have a user to update. query-before-test INSERT INTO groupdugga(hash,active_users) VALUES("hjk4ert6", 52)

Comment (b21antbr) [inline DuggaSys/tests/showDuggaservice_test.php:31]: a query-after-test is needed to remove inserted row from DB. query-after-test DELETE FROM groupdugga WHERE hash = “hjk4ert6”;

File: DuggaSys/tests/showDuggaservice_test.php [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        'service-data' => serialize(array( // Data that service needs to execute function
+            'opt' => 'SAVDU',
+            'hash' => 'ghj1jfg2'
+        )),
+        'filter-output' => serialize(array( // Filter what output to use in assert test, use none to use all ouput from service
+            'none'
+        )),
+    ),
+    'update submitted dugga' => array(
+        'expected-output' => '{"debug":"[Guest] Missing hash\/password\/variant!","param":"{}","answer":"UNK","danswer":"UNK","score":0,"highscoremode":"","grade":"UNK","submitted":"","marked":"","deadline":"UNK","release":"UNK","files":[],"userfeedback":"UNK","feedbackquestion":"UNK","variant":"UNK","ishashindb":false,"variantsize":"UNK","variantvalue":"UNK","password":"UNK","hashvariant":"UNK","isFileSubmitted":"UNK","isTeacher":0,"variants":[],"duggaTitle":"UNK","hash":"UNK","hashpwd":"UNK","opt":"UNK","link":"UNK"}',
```
Comment (b21antbr) [inline DuggaSys/tests/showDuggaservice_test.php:54]: a query-before-test is needed to be able to update a dugga. query-before-test INSERT INTO course(cid, creator) VALUES (9999, 1);

File: DuggaSys/tests/showDuggaservice_test.php [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+            'hash' => 'hjk4ert6',
+            'AUtoken' => '9999'
+        )),
+        'filter-output' => serialize(array( // Filter what output to use in assert test, use none to use all ouput from service
+            'none'
+        )),
+    ),
+    'get data from userAnswer' => array(
+        'expected-output' => '{"debug":"[Guest] Missing hash\/password\/variant!","param":"{}","answer":"UNK","danswer":"UNK","score":0,"highscoremode":"","grade":"UNK","submitted":"","marked":"","deadline":"UNK","release":"UNK","files":[],"userfeedback":"UNK","feedbackquestion":"UNK","variant":"UNK","ishashindb":false,"variantsize":"UNK","variantvalue":"UNK","password":"UNK","hashvariant":"UNK","isFileSubmitted":"UNK","isTeacher":0,"variants":[],"duggaTitle":"UNK","hash":"UNK","hashpwd":"UNK","opt":"UNK","link":"UNK"}',
+        'service' => 'localhost/LenaSYS/DuggaSys/showDuggaservice.php',
```
Comment (c21axepe) [inline DuggaSys/tests/showDuggaservice_test.php:44]: Where are the inserts and selects needed for this service?
an INSERT INTO course and INSERT INTO listentries is needed for an INSTERT INTO userAnswer.

Every insert needs to be deleted aswell dont forget that!

File: DuggaSys/tests/showDuggaservice_test.php (hunk 5 above)
Comment (b21antbr) [reply DuggaSys/tests/showDuggaservice_test.php:54]: anoither query-before-test is needed to insert listentries. INSERT INTO listentries(cid,entryname,link,kind,pos,creator,visible,vers,gradesystem,highscoremode,feedbackenabled,feedbackquestion) VALUES (9999,  "Inserttobedeleted", "UNK", 4, 12, 2, 1, 1337, 1, 1, 0, "UNK");

File: DuggaSys/tests/showDuggaservice_test.php [hunk 7]
```diff
@@ -0,0 +1,160 @@
+<?php
+
+include "../../Shared/test.php";
+
+$testsData = array(
+    'Get active users' => array(
+        'expected-output' => '{"debug":"[Guest] Missing hash\/password\/variant!","param":"{}","answer":"UNK","danswer":"UNK","score":0,"highscoremode":"","grade":"UNK","submitted":"","marked":"","deadline":"UNK","release":"UNK","files":[],"userfeedback":"UNK","feedbackquestion":"UNK","variant":"UNK","ishashindb":false,"variantsize":"UNK","variantvalue":"UNK","password":"UNK","hashvariant":"UNK","isFileSubmitted":"UNK","isTeacher":0,"variants":[],"duggaTitle":"UNK","hash":"UNK","hashpwd":"UNK","opt":"UNK","link":"UNK"}',
+        'service' => 'localhost/LenaSYS/DuggaSys/showDuggaservice.php',
```
Comment (c21axepe) [inline DuggaSys/tests/showDuggaservice_test.php:8]: ALL 'service' addresses are wrong, It should be:
https://cms.webug.se/root/G2/students/"YOUR USER"/LenaSYS/DuggaSys/showDuggaservice.php

File: DuggaSys/tests/showDuggaservice_test.php [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+            'hash' => 'hnf3j58s'
+        )),
+        'filter-output' => serialize(array( // Filter what output to use in assert test, use none to use all ouput from service
+            'debug',
+            'readonly'
+        )),
+    ),
+    'Create active users' => array(
+        'expected-output' => '{"debug":"[Guest] Missing hash\/password\/variant!","param":"{}","answer":"UNK","danswer":"UNK","score":0,"highscoremode":"","grade":"UNK","submitted":"","marked":"","deadline":"UNK","release":"UNK","files":[],"userfeedback":"UNK","feedbackquestion":"UNK","variant":"UNK","ishashindb":false,"variantsize":"UNK","variantvalue":"UNK","password":"UNK","hashvariant":"UNK","isFileSubmitted":"UNK","isTeacher":0,"variants":[],"duggaTitle":"UNK","hash":"UNK","hashpwd":"UNK","opt":"UNK","link":"UNK"}',
+        'service' => 'localhost/LenaSYS/DuggaSys/showDuggaservice.php',
```
Comment (c21axepe) [inline DuggaSys/tests/showDuggaservice_test.php:20]: Wrong address

File: DuggaSys/tests/showDuggaservice_test.php [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+            'opt' => 'SAVDU',
+            'hash' => 'ghj1jfg2'
+        )),
+        'filter-output' => serialize(array( // Filter what output to use in assert test, use none to use all ouput from service
+            'none'
+        )),
+    ),
+    'update submitted dugga' => array(
+        'expected-output' => '{"debug":"[Guest] Missing hash\/password\/variant!","param":"{}","answer":"UNK","danswer":"UNK","score":0,"highscoremode":"","grade":"UNK","submitted":"","marked":"","deadline":"UNK","release":"UNK","files":[],"userfeedback":"UNK","feedbackquestion":"UNK","variant":"UNK","ishashindb":false,"variantsize":"UNK","variantvalue":"UNK","password":"UNK","hashvariant":"UNK","isFileSubmitted":"UNK","isTeacher":0,"variants":[],"duggaTitle":"UNK","hash":"UNK","hashpwd":"UNK","opt":"UNK","link":"UNK"}',
+        'service' => 'localhost/LenaSYS/DuggaSys/showDuggaservice.php',
```
Comment (c21axepe) [inline DuggaSys/tests/showDuggaservice_test.php:55]: Wrong address

File: DuggaSys/tests/showDuggaservice_test.php [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+            'haspwd' => 'asddasdd',
+            'answer' => 'NULL'
+        )),
+        'filter-output' => serialize(array( // Filter what output to use in assert test, use none to use all ouput from service
+            'none'
+        )),
+    ),
+    'super-view data from useranswer on hash' => array(
+        'expected-output' => '{"debug":"[Guest] Missing hash\/password\/variant!","param":"{}","answer":"UNK","danswer":"UNK","score":0,"highscoremode":"","grade":"UNK","submitted":"","marked":"","deadline":"UNK","release":"UNK","files":[],"userfeedback":"UNK","feedbackquestion":"UNK","variant":"UNK","ishashindb":false,"variantsize":"UNK","variantvalue":"UNK","password":"UNK","hashvariant":"UNK","isFileSubmitted":"UNK","isTeacher":0,"variants":[],"duggaTitle":"UNK","hash":"UNK","hashpwd":"UNK","opt":"UNK","link":"UNK"}',
+        'service' => 'localhost/LenaSYS/DuggaSys/showDuggaservice.php',
```
Comment (c21axepe) [inline DuggaSys/tests/showDuggaservice_test.php:93]: Wrong address

File: DuggaSys/tests/showDuggaservice_test.php [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+        'expected-output' => '{"debug":"[Guest] Missing hash\/password\/variant!","param":"{}","answer":"UNK","danswer":"UNK","score":0,"highscoremode":"","grade":"UNK","submitted":"","marked":"","deadline":"UNK","release":"UNK","files":[],"userfeedback":"UNK","feedbackquestion":"UNK","variant":"UNK","ishashindb":false,"variantsize":"UNK","variantvalue":"UNK","password":"UNK","hashvariant":"UNK","isFileSubmitted":"UNK","isTeacher":0,"variants":[],"duggaTitle":"UNK","hash":"UNK","hashpwd":"UNK","opt":"UNK","link":"UNK"}',
+        'query-before-test-1' => "INSERT INTO course(cid, creator) VALUES (9999, 1);",
+        'query-before-test-2' => "INSERT INTO listentries(cid,entryname,link,kind,pos,creator,visible,vers,gradesystem,highscoremode,feedbackenabled,feedbackquestion) VALUES (9999, 'Inserttobedeleted', 'UNK', 4, 12, 2, 1, 1337, 1, 1, 0, 'UNK');",
+        'query-before-test-3' => "SELECT MAX(lid) FROM listentries",
+        'query-after-test-1' => "DELETE FROM listentries WHERE cid = 9999;",
+        'query-after-test-2' => "DELETE FROM course WHERE cid = 9999;",
+        'query-variables' => "moment",
+        'service' => 'localhost/LenaSYS/DuggaSys/showDuggaservice.php',
```
Comment (c21axepe) [inline DuggaSys/tests/showDuggaservice_test.php:125]: Wrong address

File: DuggaSys/tests/showDuggaservice_test.php [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+        'service-data' => serialize(array( // Data that service needs to execute function
+            'moment' => '<!query-before-test3!> <[0][listentries]>'
+        )),
+        'filter-output' => serialize(array( // Filter what output to use in assert test, use none to use all ouput from service
+            'none'
+        )),
+    ),
+    'view data on userAnswer' => array(
+        'expected-output' => '{"debug":"[Guest] Missing hash\/password\/variant!","param":"{}","answer":"UNK","danswer":"UNK","score":0,"highscoremode":"","grade":"UNK","submitted":"","marked":"","deadline":"UNK","release":"UNK","files":[],"userfeedback":"UNK","feedbackquestion":"UNK","variant":"UNK","ishashindb":false,"variantsize":"UNK","variantvalue":"UNK","password":"UNK","hashvariant":"UNK","isFileSubmitted":"UNK","isTeacher":0,"variants":[],"duggaTitle":"UNK","hash":"UNK","hashpwd":"UNK","opt":"UNK","link":"UNK"}',
+        'service' => 'localhost/LenaSYS/DuggaSys/showDuggaservice.php',
```
Comment (c21axepe) [inline DuggaSys/tests/showDuggaservice_test.php:135]: Wrong address

File: DuggaSys/tests/showDuggaservice_test.php [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+            'hash' => 'ghj1ghj2',
+            'hashpwd' => 'asddasdd'
+        )),
+        'filter-output' => serialize(array( // Filter what output to use in assert test, use none to use all ouput from service
+            'none'
+        )),
+    ),
+    'super-view data on quizname' => array(
+        'expected-output' => '{"debug":"[Guest] Missing hash\/password\/variant!","param":"{}","answer":"UNK","danswer":"UNK","score":0,"highscoremode":"","grade":"UNK","submitted":"","marked":"","deadline":"UNK","release":"UNK","files":[],"userfeedback":"UNK","feedbackquestion":"UNK","variant":"UNK","ishashindb":false,"variantsize":"UNK","variantvalue":"UNK","password":"UNK","hashvariant":"UNK","isFileSubmitted":"UNK","isTeacher":0,"variants":[],"duggaTitle":"UNK","hash":"UNK","hashpwd":"UNK","opt":"UNK","link":"UNK"}',
+        'service' => 'localhost/LenaSYS/DuggaSys/showDuggaservice.php',
```
Comment (c21axepe) [inline DuggaSys/tests/showDuggaservice_test.php:146]: Wrong address

File: DuggaSys/tests/showDuggaservice_test.php (hunk 5 above)
Comment (b21antbr) [reply DuggaSys/tests/showDuggaservice_test.php:54]: after this you need another query-before-test SELECT MAX(lid) FROM listentries
This value needs to be saved in query-variables as lid.
lid has to be stated in the send as 'lid' => <!query-before-test3!><*[0][listenentries]*>

File: DuggaSys/tests/showDuggaservice_test.php (hunk 3 above)
Comment (c21axepe) [reply DuggaSys/tests/showDuggaservice_test.php:26]: This applies to ALL other services

File: DuggaSys/tests/showDuggaservice_test.php (hunk 5 above)
Comment (b21antbr) [reply DuggaSys/tests/showDuggaservice_test.php:54]: then another query-before-test4 needs to be added(I've missed numbering all, but they start on 1 and increment for each new query, important to have these in the correct order.) query-before-test4 INSERT INTO userAnswer(cid,hash,password,moment) VALUES (9999, “dfg4zxc5”, “asdfasdf”, ?);

This questionmark is the saved value from query-variables.

File: DuggaSys/tests/showDuggaservice_test.php [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+        'query-before-test-1' => "INSERT INTO course(cid, creator) VALUES (9999, 1);",
+        'query-before-test-2' => "INSERT INTO listentries(cid,entryname,link,kind,pos,creator,visible,vers,gradesystem,highscoremode,feedbackenabled,feedbackquestion) VALUES (9999, 'Inserttobedeleted', 'UNK', 4, 12, 2, 1, 1337, 1, 1, 0, 'UNK');",
+        'query-before-test-3' => "SELECT MAX(lid) FROM listentries",
+        'query-after-test-1' => "DELETE FROM listentries WHERE cid = 9999;",
+        'query-after-test-2' => "DELETE FROM course WHERE cid = 9999;",
+        'query-variables' => "moment",
+        'service' => 'localhost/LenaSYS/DuggaSys/showDuggaservice.php',
+        'service-data' => serialize(array( // Data that service needs to execute function
+            'moment' => '<!query-before-test3!> <[0][listentries]>'
```
Comment (c21axepe) [inline DuggaSys/tests/showDuggaservice_test.php:127]: Missing * at start and end, Should be written like this:
'moment' => '<!query-before-test3!> <*[0][listentries]*>'

**Review by c21axepe — CHANGES_REQUESTED** [review]
Changes requested.

File: DuggaSys/tests/showDuggaservice_test.php [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+        )),
+        'filter-output' => serialize(array( // Filter what output to use in assert test, use none to use all ouput from service
+            'none'
+        )),
+    ),
+    'view data on userAnswer' => array(
+        'expected-output' => '{"debug":"[Guest] Missing hash\/password\/variant!","param":"{}","answer":"UNK","danswer":"UNK","score":0,"highscoremode":"","grade":"UNK","submitted":"","marked":"","deadline":"UNK","release":"UNK","files":[],"userfeedback":"UNK","feedbackquestion":"UNK","variant":"UNK","ishashindb":false,"variantsize":"UNK","variantvalue":"UNK","password":"UNK","hashvariant":"UNK","isFileSubmitted":"UNK","isTeacher":0,"variants":[],"duggaTitle":"UNK","hash":"UNK","hashpwd":"UNK","opt":"UNK","link":"UNK"}',
```
Comment (b21antbr) [inline DuggaSys/tests/showDuggaservice_test.php:134]: query-before-test needed for INSERT INTO course(cid, creator) VALUES (9999, 1);

Comment (b21antbr) [reply DuggaSys/tests/showDuggaservice_test.php:134]: query-before-test2 needed for INSERT INTO listentries(cid,entryname,link,kind,pos,creator,visible,vers,gradesystem,highscoremode,feedbackenabled,feedbackquestion) VALUES (9999, "Inserttobedeleted", "UNK", 4, 12, 2, 1, 1337, 1, 1, 0, "UNK");

Comment (b21antbr) [reply DuggaSys/tests/showDuggaservice_test.php:134]: another query-before-test3 needed for SELECT MAX(lid) FROM listentries, this value has to be saved in query-variables as moment

Comment (b21antbr) [reply DuggaSys/tests/showDuggaservice_test.php:134]: another query-before-test4 needed for INSERT INTO userAnswer (cid,quiz,moment,vers,variant,hash,password, useranswer) VALUES(9999, 52432, 1, ? , 3, “ghj1ghj2” , “asddasdd”,NULL)

NOTE: this query is completely wrong in the test-case file, make sure these values are correct.

**Review by b21antbr — CHANGES_REQUESTED** [review]
changes requested as commented in the code.

File: DuggaSys/tests/showDuggaservice_test.php [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+        'query-before-test-1' => "INSERT INTO course(cid, creator) VALUES (9999, 1);",
+        'query-before-test-2' => "INSERT INTO listentries(cid,entryname,link,kind,pos,creator,visible,vers,gradesystem,highscoremode,feedbackenabled,feedbackquestion) VALUES (9999, 'Inserttobedeleted', 'UNK', 4, 12, 2, 1, 1337, 1, 1, 0, 'UNK');",
+        'query-before-test-3' => "SELECT MAX(lid) FROM listentries",
+        'query-after-test-1' => "DELETE FROM userAnswer ORDER BY aid DESC LIMIT 1;",
+        'query-after-test-2' => "DELETE FROM listentries WHERE cid = 9999;",
+        'query-after-test-3' => "DELETE FROM course WHERE cid = 9999;",
+        'query-variables' => "moment",
+        'service' => 'localhost/LenaSYS/DuggaSys/showDuggaservice.php',
```
[the remark this answers is not in the document]
Comment (b21antbr) [reply DuggaSys/tests/showDuggaservice_test.php:75]: querys before and after test seem to be correct here, follow these for the other comments.

**Comment by d21erisu** [comment]
Closing since its time for merge. A new issue-branch and PR needs to be made for this content to be merged.

**Pull request closed without merging**